python/grass/script/db.py

```diff
@@ -233,23 +246,139 @@ def db_table_in_vector(table, mapset=".", env=None):
+        self._pdriver = self._c_funcs["db_start_driver_open_database"](
+            self._driver_name,
+            self._c_funcs["Vect_subst_var"](self._database, byref(map)),
+        )
+        if not self._pdriver:
+            fatal(
+                _("Unable to open database <{db}> by driver <{driver}>.").format(
+                    db=self._database, driver=self._driver_name
+                )
+            )
 
-    :return: SQL command as string
-    """
-    if driver in ("sqlite", "pg", "mysql"):
-        return "COMMIT"
-    return ""
+    def _begin_transaction(self):
+        """Begin DB transaction."""
+        self._c_funcs["G_gisinit"]("")
+
+        self._init_driver()
+
+        ret = self._c_funcs["db_begin_transaction"](self._pdriver)
+        if ret != self._c_funcs["DB_OK"]:
+            self._c_funcs["db_close_database_shutdown_driver"](self._pdriver)
+            fatal(
+                _(
+                    "Error while start database <{db}> transaction by"
+                    " driver <{driver}>."
+                ).format(db=self._database, driver=self._driver_name)
+            )
+
+    def _commit_transaction(self):
+        """Commit DB transaction."""
+        ret = self._c_funcs["db_commit_transaction"](self._pdriver)
+        if ret != self._c_funcs["DB_OK"]:
+            self._c_funcs["db_close_database_shutdown_driver"](self._pdriver)
+            fatal(
+                _(
+                    "Error while commit database <{db}> transaction"
+                    " by driver <{driver}>."
+                ).format(db=self._database, driver=self._driver_name)
+            )
+        self._c_funcs["db_close_database_shutdown_driver"](self._pdriver)
+
+    def _execute(self, sql):
+        """Execute SQL
+
+        :param str|list|tuple sql: SQL command string or list of SQLs
+                                   commands
+        """
+        stmt = self._c_funcs["dbString"]()
+        self._c_funcs["db_init_string"](byref(stmt))
+        self._c_funcs["db_set_string"](byref(stmt), sql)
+        if (
+            self._c_funcs["db_execute_immediate"](self._pdriver, byref(stmt))
+            != self._c_funcs["DB_OK"]
+        ):
+            self._c_funcs["db_free_string"](byref(stmt))
+            self._c_funcs["db_close_database_shutdown_driver"](self._pdriver)
+            fatal(_("Error while executing SQL <{}>.").format(sql))
+        self._c_funcs["db_free_string"](byref(stmt))
+
+    def execute(self, sql):
+        """Execute SQL
+
+        :param str|list|tuple sql: SQL command string or list of SQLs
+                                   statement
+        """
+        # Begin DB transaction
+        self._begin_transaction()
+        # Execute SQL string
+        if isinstance(sql, (list, tuple)):
+            for statement in sql:
+                self._execute(sql=statement)
+        else:
+            self._execute(sql)
+        # Commit DB transaction
+        self._commit_transaction()
```

Comment:
While consistent within this code, database driver start and shutdown within one function are perhaps little unexpected. One probably should not do more than couple execute calls per process otherwise it might be a performance hit which is a big reason to use transactions in our case. This needs clear documentation.